src/trace/layers/ordered/mod.rs

```diff
@@ -298,6 +294,111 @@ where
+    {
+        match trie1.keys[*lower1].cmp(&trie2.keys[*lower2]) {
+            Ordering::Less => {
+                // determine how far we can advance lower1 until we reach/pass lower2
+                let step = 1 + advance(&trie1.keys[(1 + *lower1)..upper1], |x| {
+                    x < &trie2.keys[*lower2]
+                });
+                let step = min(step, 1_000);
+                self.copy_range(trie1, *lower1, *lower1 + step);
+                *lower1 += step;
+            }
+
+            Ordering::Equal => {
+                let lower = self.vals.boundary();
+                // record vals_length so we can tell if anything was pushed.
+                let upper = self.vals.push_merge(
+                    trie1.vals.cursor_from(
+                        trie1.offs[*lower1].into_usize(),
+                        trie1.offs[*lower1 + 1].into_usize(),
+                    ),
+                    trie2.vals.cursor_from(
+                        trie2.offs[*lower2].into_usize(),
+                        trie2.offs[*lower2 + 1].into_usize(),
+                    ),
+                );
+                if upper > lower {
+                    self.keys.push(trie1.keys[*lower1].clone());
+                    self.offs.push(O::from_usize(upper));
+                }
+
+                *lower1 += 1;
+                *lower2 += 1;
+            }
+
+            Ordering::Greater => {
+                // determine how far we can advance lower2 until we reach/pass lower1
+                let step = 1 + advance(&trie2.keys[(1 + *lower2)..upper2], |x| {
+                    x < &trie1.keys[*lower1]
+                });
+                let step = min(step, 1_000);
+                self.copy_range(trie2, *lower2, *lower2 + step);
+                *lower2 += step;
+            }
+        }
+    }
+
+    /// Push a key and all of its associated values to the current builder
+    ///
+    /// Can be more efficient than repeatedly calling `.push_tuple()` because it
+    /// doesn't require an owned key for every value+diff pair which can allow
+    /// eliding unnecessary clones
+    pub fn with_key<F>(&mut self, key: K, with: F)
+    where
+        K: Eq,
+        L: TupleBuilder,
+        O: OrdOffset,
+        F: for<'a> FnOnce(OrderedBuilderVals<'a, K, L, O>),
+    {
+        let mut pushes = 0;
+        let vals = OrderedBuilderVals {
+            builder: self,
+            pushes: &mut pushes,
+        };
+        with(vals);
+
+        // If the user's closure actually added any elements, push the key
+        if pushes != 0
+            && (self.keys.is_empty()
+                || !self.offs[self.keys.len()].is_zero()
+                || self.keys[self.keys.len() - 1] != key)
+        {
+            if !self.keys.is_empty() && self.offs[self.keys.len()].is_zero() {
+                self.offs[self.keys.len()] = O::from_usize(self.vals.boundary());
+            }
+            self.keys.push(key);
+            self.offs.push(O::zero());
```

Comment:
Since we're leaving the key in a potentially unfinished state, it just means that you can call `builder.with_key(1, ...); builder.push((1, 10));` and it'll function correctly. It keeps this function in line with `OrderedBuilder::push()`